<div style="border:solid Chocolate 2px; padding: 40px">

<b>Олег, приветствую!</b>

Буду рада с тобой познакомиться. Меня зовут Евгения Маракова и я буду ревьюером твоего проекта. Если ты не против, то предлагаю общаться на "ты":) Если удобнее на "вы", то нет проблем, только скажи об этом.

В процессе работы ты встретишь комментарии <font color='green'>зеленого</font>, <font color='gold'>желтого</font> и <font color='red'>красного</font> цветов. Что они значат, смотри ниже:
    <br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Решение на этом шаге является полностью правильным.
</div>
    

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Решение на этом шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
 Решение на этом шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
Увидев мой комментарий, не удаляй его, он будет очень полезен в случае повторной проверки работы😊
    
Свой комментарий помечай так, как показано ниже, тогда нам будет легче вести диалог, а мне выполнить повторную проверку проекта.
    <div class="alert alert-info"> <b>Комментарий студента: </b> </div>
    


# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [29]:
data_1 = pd.read_csv('/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

In [23]:
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [24]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#Создадим сводную таблицу, где соберем средние значения столбца 'debt' по исследуемой группе, 
# посчитаем количество записей в каждой и найдем количество должников
children_correlation = data.pivot_table(index = ['children'], 
                                        values='debt', 
                                        aggfunc = ('sum', 'mean', lambda X: X.count()))

# Создадим функцию, форматирующую данные к процентному формату                                                                      
def percent(x):
    return "{0:.2%}".format(x)

# Применим функцию к сводной таблице для столбца 'mean'
children_correlation['mean'] = children_correlation['mean'].apply(percent) 
 
# Сортировка    
children_correlation = children_correlation.sort_values(by = 'mean', ascending = False)
# Переименуем столбцы для удобства
children_correlation = children_correlation.rename(columns = {'<lambda_0>': 'Количество кредитополучателей',
                                                              'sum': 'Количество должников', 'mean': 'Доля должников'})
                                                              
                                                              
children_correlation.reindex(columns = ['Количество кредитополучателей','Количество должников', 'Доля должников'])



Количество кредитополучателей  Количество должников Доля должников
children                                                                    
4                                  41.0                   4.0          9.76%
2                                2052.0                 194.0          9.45%
1                                4808.0                 444.0          9.23%
3                                 330.0                  27.0          8.18%
0                               14091.0                1063.0          7.54%
5                                   9.0                   0.0          0.00%

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
 Олег, для наглядности, необходимо вывести сводную таблицу: категория/всего кредитополучателей/всего должников/доля должников. У тебя не хватает должников:) 
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> +

</div>

**Вывод:** 
Наиболее ответственной группой являются люди без детей (7,54%), а самыми безответственными группа людей с 4 детьми (9,76%). ПО группе 5 выводы делать не представляеся возможным, поскольку в наблюдение попало всего 9 записей.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Верный вывод!
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#Создадим сводную таблицу, где соберем средние значения столбца 'debt' по исследуемой группе, 
# посчитаем количество записей в каждой и найдем количество должников
family_status_correlation = data.pivot_table(index = ['family_status'], 
                                        values='debt', 
                                        aggfunc = ('sum', 'mean', lambda X: X.count()))

# Создадим функцию, форматирующую данные к процентному формату                                                                      
def percent(x):
    return "{0:.2%}".format(x)

# Применим функцию к сводной таблице для столбца 'mean'
family_status_correlation['mean'] = family_status_correlation['mean'].apply(percent) 
 
# Сортировка    
family_status_correlation = family_status_correlation.sort_values(by = 'mean', ascending = False)
# Переименуем столбцы для удобства
family_status_correlation = family_status_correlation.rename(columns = {'<lambda_0>': 'Количество кредитополучателей',
                                                              'sum': 'Количество должников', 'mean': 'Доля должников'})
                                                              
                                                              
family_status_correlation.reindex(columns = ['Количество кредитополучателей','Количество должников', 'Доля должников'])

Количество кредитополучателей  Количество должников  \
family_status                                                                
Не женат / не замужем                         2796.0                 273.0   
гражданский брак                              4134.0                 385.0   
женат / замужем                              12261.0                 927.0   
в разводе                                     1189.0                  84.0   
вдовец / вдова                                 951.0                  63.0   

                      Доля должников  
family_status                         
Не женат / не замужем          9.76%  
гражданский брак               9.31%  
женат / замужем                7.56%  
в разводе                      7.06%  
вдовец / вдова                 6.62%

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
ПО таблице те же замечания:)
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> +

</div>

**Вывод:** Наиболее ответственной группой данных является "вдовец / вдова" (6,62%), а самой безответственной "Не женат / не замужем" (9.76%)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Отлично!
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#Создадим сводную таблицу, где соберем средние значения столбца 'debt' по исследуемой группе, 
# посчитаем количество записей в каждой и найдем количество должников
total_income_category_correlation = data.pivot_table(index = ['total_income_category'], 
                                        values='debt', 
                                        aggfunc = ('sum', 'mean', lambda X: X.count()))

# Создадим функцию, форматирующую данные к процентному формату                                                                      
def percent(x):
    return "{0:.2%}".format(x)

# Применим функцию к сводной таблице для столбца 'mean'
total_income_category_correlation['mean'] = total_income_category_correlation['mean'].apply(percent) 
 
# Сортировка    
total_income_category_correlation = total_income_category_correlation.sort_values(by = '<lambda_0>', ascending = False)
# Переименуем столбцы для удобства
total_income_category_correlation = total_income_category_correlation.rename(columns = {'<lambda_0>': 'Количество кредитополучателей',
                                                              'sum': 'Количество должников', 'mean': 'Доля должников'})
                                                              
                                                              
total_income_category_correlation.reindex(columns = ['Количество кредитополучателей','Количество должников', 'Доля должников'])

# Разбивка людей по группам доходов:
#        if 0 <= income <= 30000:
#            return 'E'
#        elif 30001 <= income <= 50000:
#            return 'D'
#        elif 50001 <= income <= 200000:
#            return 'C'
#        elif 200001 <= income <= 1000000:
#            return 'B'
#        elif income >= 1000001:
#            return 'A'

Количество кредитополучателей  Количество должников  \
total_income_category                                                        
C                                            15921.0                1353.0   
B                                             5014.0                 354.0   
D                                              349.0                  21.0   
A                                               25.0                   2.0   
E                                               22.0                   2.0   

                      Доля должников  
total_income_category                 
C                              8.50%  
B                              7.06%  
D                              6.02%  
A                              8.00%  
E                              9.09%

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
 Таблица.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> +

</div>

**Вывод:** Наиболее ответственно к выплате кредитов в срок подходит группа "D", имеющая средний доход в диапазоне от 30000 до 50000 единиц. Наименее ответственно к выплатам в срок подходит группа "Е" cо средним значением дохода не более 30000

<div class="alert alert-info"> <b>Комментарий студента: </b> </div>
**Вывод №2** Исходя из имеющихся данных, наиболее широкой категорией является "С", что логично, поскольку данная категория является "средним классом". На основе этого закономерно и то, что доля должников будет достаточно высока.
В группе "B", с увеличением класса доходов, закономерно уменьшается и среднее количество должников, однако данная логика нарушается, когда мы анализируем класс с наивысшим показателем доходов "А". Причиной этого может быть то, что чем больше доход, тем больше и кредит, а соответственно выше ставки и рисковее цели.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Обрати внимание на несбалансированность выборок. Рекомендую, для выявления зависимости, сравнить категории с наибольшими показателями - В и С.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Хорошо!

</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [27]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#Создадим сводную таблицу, где соберем средние значения столбца 'debt' по исследуемой группе, 
# посчитаем количество записей в каждой и найдем количество должников
purpose_category_correlation = data.pivot_table(index = ['purpose_category'], 
                                        values='debt', 
                                        aggfunc = ('sum', 'mean', lambda X: X.count()))

# Создадим функцию, форматирующую данные к процентному формату                                                                      
def percent(x):
    return "{0:.2%}".format(x)

# Применим функцию к сводной таблице для столбца 'mean'
purpose_category_correlation['mean'] = purpose_category_correlation['mean'].apply(percent) 
 
# Сортировка    
purpose_category_correlation = purpose_category_correlation.sort_values(by = 'mean', ascending = False)
# Переименуем столбцы для удобства
purpose_category_correlation = purpose_category_correlation.rename(columns = {'<lambda_0>': 'Количество кредитополучателей',
                                                              'sum': 'Количество должников', 'mean': 'Доля должников'})
                                                              
                                                              
purpose_category_correlation.reindex(columns = ['Количество кредитополучателей','Количество должников', 'Доля должников'])

Количество кредитополучателей  Количество должников  \
purpose_category                                                                
операции с автомобилем                           4279.0                 400.0   
получение образования                            3988.0                 369.0   
проведение свадьбы                               2313.0                 183.0   
операции с недвижимостью                        10751.0                 780.0   

                         Доля должников  
purpose_category                         
операции с автомобилем            9.35%  
получение образования             9.25%  
проведение свадьбы                7.91%  
операции с недвижимостью          7.26%

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Таблица.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> +

</div>

**Вывод:** Наиболее ответственная группа людей, которая берет кредит(ипотеку) на операции с недвижимостью (7,26%), а наименее ответственно для операций с автомобилем (9,35%)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Хорошо! 
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски коснулись столбцов с количеством отработанных дней и доходом. Данные столбцы имеют одинковое количество пропусков (2174) и скорее всего взаимосвязаны. Возможно не указав стаж работы ты не можешь указать и средний доход, на уровне раполнения формы подачи данных на получение кредита. Аналогично этому, причиной существования пропусков в этих столбцах может быть связана с тем, что человек  мог не в поступать на официальную работу (работать в серую), а соответственно не вступая, официально, в трудовую деятельность, ты не можешь иметь и доход (или не желать указывать свой доход, во втором случае).

Поиск иных причин может ограничиваться лишь воображением, для однозначного ответа недостаточно данных.

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Ты верно указал примеры. Также нужно назвать два фактора, к которым относятся данные примеры.
</div>

<div class="alert alert-info"> <b>Комментарий студента: </b> </div>
В целом не совсем понимаю как понять фактор, к которому относится мой пример. 

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Нужно было указать человеческий фактор и техническую ошибку:)

</div>

In [33]:
filtered_df = data_1[data_1['days_employed'].isnull()]
filtered_df
# А если быть точнее, то не вижу закономерностей при появлении данных.. максимум что еще могу выжать, так это банальные ошибки 
# при вводе данных

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Покольку поиск средней, зачастую не отражает действительности. В случае анализа группы людей, где зарплаты не имеют больших разбросов поиск средней оправдан, однако на реальных данных наиболее логичным является поиск "середины"

Пример! 
Данные о зарплатах: 30к, 50к, 75к, 160к, 1,5кк
Поиск средней на основе придуманных зарплат: средние зарплаты = 353 тысячирублей
Поиск медианы на основе придуманныз зарплат: медианная зарплата составляет 75 тысяч рублей
Вывод в выводе: логичнее медиана!

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Согласна, медиана менее чувствительна к выбросам!
</div>

<div class="alert alert-info"> <b>Комментарий студента: </b> </div>

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

В основе исследования лежали данные кредитной организации, содержащие достаточно широкий перечень данных о клиентах:
    1)children - количество детей 
    2)days_employed - стаж работы
    3)dob_years - количество лет 
    4)education - уровень образования
    5)education_id - id уровня образования
    6)family_status - семейное положение      
    7)family_status_id - id статуса семейного положения
    8)gender - пол
    9)income_type - трудовой статус         
    10)debt - имел ли непогашенные в срок задолженности                
    11)total_income - доход       
    12)purpose - цель кредита
Также были добавлены 2 категориальных поля: 
    1) total_income_category - уровень доходов с привязкой к классам от 'A' до "D"
    2) purpose_category - категория причины взятия кредита
При работе с данными были выявлены следующие проблемы: 
    1) столбец 'days_employed' содержал пропуски и отрицательные значения
    2) столбец 'total_income' содержал пропуски в значениях
    3) Датасет содержал выбросы в столбце 'children' в виде значений = 20, -1
    4) столбец 'education' содержит одинаковые значения, записанные в разных регистрах


**Рекомендации**:
    1) поставить запред на ввод значений для раздела "дети" на отрицательные значения, чтобы не хранить в базе данных заведомо лишние значения
    2) поставить запред на ввод значений для разделов "доход" и "стаж" на отрицательные значения, чтобы не тратить время на предобработку данных
    3) При ведении базы данныхнеобходимо приводить к нижнему регистру поле "образование" и иные поля, содержащие данные о **типе** пользователя

При выдаче кредитов, наиболее привлекательными являются клиенты, имеющие следующие показатели: 
Наиболее выгодным для кредитной организации являются:
    1) люди с доходом в диапазоне от 30 000 до 50 000 тысяч ед.
    2) люди приобритающие недвижимость 
    3) без детей
    4) вдовы/вдовцы
...
(и так далее по порядку)
...
Наиболее рисковыми являются люди которые: 
    1) имеют доход ниже 30 000 тыс ед.
    2) берущие кредит на операции с автомобилем
    3) не втупавшие в брак
    4) имеющие детей в целом, но наиболее рисковые - 4 ребенка

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
<b>Необходимо исправить❌:</b>
Олег, общий вывод - обязательный и важный элемент работы. Его нужно сформулировать, взяв основные результаты из всех шагов проекта. Начать нужно с того, с какими данными мы работали и какие действия провели на этапе предобработки, а закончить выявленными при ответе на вопросы зависимостями и рекомендациями заказчику по улучшению сбора данных и скоринга. Заказчик, зачастую, читает только общий вывод исследования - поэтому он должен быть максимально информативным. 
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Олег, отлично! Работа принята, хорошего дня:)

</div>

<div style="border:solid Chocolate 2px; padding: 40px">
Олег, давай подведем итоги. Ты большой молодец, что старался и проделал такую большую работу. Шаг 1 и 2 ты выполнил без проблем, что очень хорошо:)
    
    Разберем над чем стоит поработать:

    1. Необходимо выводить сводные таблицы со всеми столбцами сразу, а не по отдельности. Для этого можно использовать метод groupby() или pivot_table().
    
    2. Нужно учитывать несбалансированность выборок.
    
    3. Указать два фактора, влияющие на появление пропусков.
    
    4. И более детально раскрыть общий вывод.
    
    Пересмотри работу и исправь ошибки. Обязательно пиши, если я могу тебе еще чем-то помочь. Удачи!)
  